### DISCUSSION TITLE
Exception in createStorageKey function due to settings contextPath being null

### DISCUSSION THREAD

**Discussion by aheitkamp** [primefaces/community#1013] [PrimeFaces] [answered]
Our application is using the PanelMenu component with collapsible menu items.  The menu is built from a MenuModel.

The failing function createStorageKey, defined in core.js:

           createStorageKey: function(b, d, e) {
                return (**c.settings.contextPath**.replace(/\//g, "-") + (e ? "" : c.settings.viewId.replace(/\//g, "-")) + b + "-" + d).toLowerCase()
            },
            
The settings collection is empty resulting in the exception when replace is called.  The createStorageKey function is called from the PanelQuery init fuction. 

Why would the settings collection be empty? Any help resolving this issue would be greatly appreciated.

Primefaces version:  13.0.3
JSF version: 2.2

Stack Trace from browser Dev Tools:

jquery.js.xhtml?ln=primefaces&v=13.0.3&e=13.0.3:2  Uncaught TypeError: Cannot read properties of undefined (reading 'replace')
    at Object.createStorageKey (core.js.xhtml?ln=primefaces&v=13.0.3&e=13.0.3:11:235)
    at d.createStorageKey (components.js.xhtml?ln=primefaces&v=13.0.3&e=13.0.3:638:188)
    at d.init (components.js.xhtml?ln=primefaces&v=13.0.3&e=13.0.3:632:268)
    at d.init (core.js.xhtml?ln=primefaces&v=13.0.3&e=13.0.3:96:383)

**Comment by melloware** [comment]
Hmmm can you reproduce it with  an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  

I have never seen that value be NULL and you can see for yourself on the showcase: 4

When I look in view source I see this.

```js
if(pf){pf.settings.locale='en_US';pf.settings.viewId='/ui/menu/panelMenu.xhtml';pf.settings.contextPath='/showcase';pf.settings.cookiesSecure=false;pf.settings.validateEmptyFields=true;
```

specifically `pf.settings.contextPath='/showcase';` so not sure how yours is NULL it should at least be `"/"`?

**Reply by tandraschko** [reply]
can you post the content of your generated head dom?

**Reply by aheitkamp** [reply]
I assume you are asking for the head element.  I have masked some information that by company policy, I cannot post online.

```xml
<head>
	<meta http-equiv="X-UA-Compatible" content="IE=Edge">
	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
	<meta http-equiv="pragma" content="no-cache">
	<meta http-equiv="cache-control" content="no-cache">
	<meta http-equiv="expires" content="0">
	<title>xxxx</title>
	<link rel="stylesheet" href="/xxx/javax.faces.resource/css/printdefault.css.xhtml" media="print">
	<link rel="stylesheet" type="text/css"
		href="/xxx/javax.faces.resource/theme.css.xhtml?ln=primefaces-xxx&amp;v=13.0.3&amp;e=13.0.3">
	<link rel="stylesheet" type="text/css" href="/xxx/javax.faces.resource/pf-custom.css.xhtml?ln=css">
	<link rel="stylesheet" type="text/css" href="/xxx/javax.faces.resource/layout.css.xhtml?ln=css">
	<link rel="stylesheet" type="text/css" href="/xxx/javax.faces.resource/xxx.css.xhtml?ln=css">
	<link rel="stylesheet" type="text/css" href="/xxx/javax.faces.resource/printdefault.css.xhtml?ln=css">
	<link rel="stylesheet" type="text/css"
		href="/xxx/javax.faces.resource/components.css.xhtml?ln=primefaces&amp;v=13.0.3&amp;e=13.0.3">
	<script type="text/javascript"
		src="/xxx/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&amp;v=13.0.3&amp;e=13.0.3"></script>
	<script type="text/javascript"
		src="/xxx/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&amp;v=13.0.3&amp;e=13.0.3"></script>
	<script type="text/javascript"
		src="/xxx/javax.faces.resource/core.js.xhtml?ln=primefaces&amp;v=13.0.3&amp;e=13.0.3"></script>
	<script type="text/javascript"
		src="/xxx/javax.faces.resource/components.js.xhtml?ln=primefaces&amp;v=13.0.3&amp;e=13.0.3"></script>
	<script type="text/javascript" src="/xxx/javax.faces.resource/common.js.xhtml?ln=js"></script>
</head>
```

**Reply by tandraschko** [reply, marked as answer]
Do you have a custom HeadRenderer? It misses all JS settings: https://github.com/primefaces/primefaces/blob/master/primefaces/src/main/java/org/primefaces/renderkit/HeadRenderer.java#L202
Pls debug.

**Reply by aheitkamp** [reply]
Turns out we do have a different HeadRender configured and that is the cause of this issue.  I will need to determine the need for our custom renderer, but using the PrimeFaces default HeadRender fixed this issue.

Thanks for pointing me in the right direction!

**Reply by melloware** [reply]
If you still need your Custom HeadRenderer you will need to copy the code from the PrimeFaces HeadRenderer to make yours work.
